Add D2k building upgrades

penev92 opened the pull request:
This PR supersedes #7597, including an update for removed traits and a couple of fixes.
As stated on the previous PR, this fixes the tech tree + some actor namings.

Same rules apply: if you can't check the original game, consult [this for buildings](http://dune2k.com/Duniverse/Games/Dune2000/Details/HierarchyCharts) and [this for units](http://dune2k.com/View/Images/Duniverse/Games/Dune2000/HierarchyCharts/Dune2000Hierarchy.jpg).
Production tab hotkeys have been fixed.

Closes #3211.
Fixes #8010.

OmegaBolt committed:
Add Upgrade production queue to D2k

OmegaBolt committed:
Add D2k building upgrades

OmegaBolt committed:
Fix D2k tech tree

- Added upgrades for the Con Yard, Barracks, Light, Heavy and High Tech Factory (last one is Atreides only)
- Updated the prerequisites of all structures, infantry and vehicles to use the upgrades
- Fixed the Ix Research building to use vanilla Dune 2000 prerequisites
- Renamed High Tech Facility to High Tech Factory like vanilla Dune 2000
- Given the Air Strike power to the High Tech Factory Upgrade like vanilla

OmegaBolt commented:
Did a quick test and it seems fine. I would just recommend swapping the Aircraft (U) and Heavy Factory (I) hotkeys so they match the order of the tabs on the sidebar at least.

pchote commented on mods/d2k/rules/infantry.yaml:
```diff
@@ -24,12 +24,40 @@ rifle:
 	AttractsWorms:
 		Intensity: 120
 
+bazooka:
```
Could you please move this back to its original location so that the diff shows what has actually changed?

pchote commented on mods/d2k/rules/infantry.yaml:
```diff
@@ -113,6 +113,40 @@ medic:
 	AttractsWorms:
 		Intensity: 180
 
+grenadier:
```
Could you please move this back to its original location so that the diff shows what has actually changed?

penev92 commented:
Reverted the infantry definition reordering. There will be time for that once we import the original actors.

pchote commented on mods/d2k/rules/player.yaml:
```diff
@@ -6,6 +6,14 @@ Player:
 		QueuedAudio: Building
 		ReadyAudio: BuildingReady
 		BlockedAudio: NoRoom
+		RequireOwner: false
```
`RequireOwner` has been removed.

pchote commented on mods/d2k/rules/structures.yaml:
```diff
@@ -757,7 +778,7 @@ palace:
 	Tooltip:
 		Name: Palace
-		Description: Unlocks elite infantry and support powers
+		Description: Unlocks elite infantry and superweapons
```
This appears to only be true for Harkonnen now.

pchote commented:
Fixes #8010.

penev92 commented:
Yeah, right, I didn't know we had an issue open about it.

MustaphaTR commented on mods/d2k/rules/structures.yaml:
```diff
@@ -694,17 +702,30 @@ hightech:
+		Prerequisite: hightech.atreides
+		Race: atreides
+	ProvidesPrerequisite@buildingname:
+	AirstrikePower:
```
`SquadSize: 3` is missing.

penev92 replied:
Good catch ;)

penev92 commented:
Also blocking this until tooltips become upgradable so I can deal with the Palace tooltips properly.

penev92 commented:
Removed `RequireOwner`, changed the palace description, added `SquadSize: 3` to the airstrike again and fixed a couple of prerequisite quirks with the airstrike power.

obrakmann commented:
This works nicely. The only thing I noticed is that the AI doesn't build upgrades. I think that should be fixed.

obrakmann commented:
This makes the AI aware of the upgrades:
```diff
diff --git a/mods/d2k/rules/ai.yaml b/mods/d2k/rules/ai.yaml
index 26e011c..0b3c03b 100644
--- a/mods/d2k/rules/ai.yaml
+++ b/mods/d2k/rules/ai.yaml
@@ -1,6 +1,7 @@
 Player:
        HackyAI@Omnius:
                Name: Omnius
+               BuildingQueues: Building, Upgrade
                UnitQueues: Infantry, Vehicle, Armor, Starport
                BuildingCommonNames:
                        ConstructionYard: conyard
@@ -21,6 +22,10 @@ Player:
                        radar: 1
                        hightech: 1
                        palace: 1
+                       upgrade.barracks: 1
+                       upgrade.light: 1
+                       upgrade.heavy: 1
+                       upgrade.hightech: 1
                BuildingFractions:
                        refinery: 20.1%
                        barracks: 0.1%
@@ -34,6 +39,10 @@ Player:
                        guntower: 8%
                        rockettower: 6%
                        power: 10%
+                       upgrade.barracks: 0.1%
+                       upgrade.light: 0.1%
+                       upgrade.heavy: 0.1%
+                       upgrade.hightech: 0.1%
                UnitsToBuild:
                        carryall: 1%
                        rifle: 6%
@@ -118,6 +127,7 @@ Player:
                                CheckRadius: 7c0
        HackyAI@Vidius:
                Name: Vidious
+               BuildingQueues: Building, Upgrade
                UnitQueues: Infantry, Vehicle, Armor, Starport
                BuildingCommonNames:
                        ConstructionYard: conyard
@@ -138,6 +148,10 @@ Player:
                        radar: 1
                        hightech: 1
                        palace: 1
+                       upgrade.barracks: 1
+                       upgrade.light: 1
+                       upgrade.heavy: 1
+                       upgrade.hightech: 1
                BuildingFractions:
                        refinery: 20.1%
                        barracks: 0.1%
@@ -151,6 +165,10 @@ Player:
                        guntower: 5%
                        rockettower: 10%
                        power: 12%
+                       upgrade.barracks: 0.1%
+                       upgrade.light: 0.1%
+                       upgrade.heavy: 0.1%
+                       upgrade.hightech: 0.1%
                UnitsToBuild:
                        carryall: 1%
                        rifle: 2%
@@ -235,6 +253,7 @@ Player:
                                CheckRadius: 7c0
        HackyAI@Gladius:
                Name: Gladius
+               BuildingQueues: Building, Upgrade
                UnitQueues: Infantry, Vehicle, Armor, Starport
                BuildingCommonNames:
                        ConstructionYard: conyard
@@ -255,6 +274,10 @@ Player:
                        radar: 1
                        hightech: 1
                        palace: 1
+                       upgrade.barracks: 1
+                       upgrade.light: 1
+                       upgrade.heavy: 1
+                       upgrade.hightech: 1
                BuildingFractions:
                        refinery: 20.1%
                        barracks: 0.1%
@@ -268,6 +291,10 @@ Player:
                        guntower: 4%
                        rockettower: 12%
                        power: 10%
+                       upgrade.barracks: 0.1%
+                       upgrade.light: 0.1%
+                       upgrade.heavy: 0.1%
+                       upgrade.hightech: 0.1%
                UnitsToBuild:
                        carryall: 1%
                        rifle: 15%
```

obrakmann committed:
Make d2k AI aware of buildings upgrades

obrakmann commented:
Nicely done. I like. :+1:

Mailaender commented:
:+1: Works as promised. Thanks! Another missing piece for the Dune 2000 mod.

penev92 commented:
This definitely deserves a [changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/5cab437a66e67abbd11a53c8100fa0d5c250a500...d1ca323cc4b62335682f0b57f8420d1c0d828904).
